experimental/collator/src/lib.rs

```diff
@@ -0,0 +1,4649 @@
+    #[inline(always)]
+    fn default() -> Self {
+        NonPrimary(0x01000100) // Low 32 bits of NO_CE
+    }
+}
+
+/// Pack a `char` and a `CanonicalCombiningClass` in
+/// 32 bits. The latter is initialized to 0xFF upon
+/// creation and can be set by calling `set_ccc`.
+/// This type is intentionally non-`Copy` to get
+/// compiler help in making sure that the class is
+/// set on the instance on which it is intended to
+/// be set and not on a temporary copy.
+///
+/// XXX check that 0xFF is actually reserved by the spec.
+#[derive(Debug)]
+struct CharacterAndClass(u32);
+
+impl CharacterAndClass {
+    pub fn new(c: char) -> Self {
+        // Setting the placeholder combining class to 0xFF to
+        // make it greater than zero when there is only one
+        // combining character and we don't do the trie lookup.
+        CharacterAndClass(u32::from(c) | (0xFF << 24))
+    }
+    pub fn character(&self) -> char {
+        unsafe { char::from_u32_unchecked(self.0 & 0xFFFFFF) }
+    }
+    pub fn ccc(&self) -> CanonicalCombiningClass {
+        CanonicalCombiningClass((self.0 >> 24) as u8)
+    }
+    // XXX need better naming here.
+    pub fn set_ccc(&mut self, ccc: &CodePointTrie<CanonicalCombiningClass>) {
+        debug_assert_eq!(self.0 >> 24, 0xFF, "This method has already been called!");
+        let scalar = self.0 & 0xFFFFFF;
+        self.0 = ((ccc.get(scalar).0 as u32) << 24) | scalar;
+    }
+}
+
+// This trivial function exists as a borrow check helper.
+#[inline(always)]
+fn sort_slice_by_ccc<'data>(
+    slice: &mut [char],
+    ccc: &CodePointTrie<'data, CanonicalCombiningClass>,
+) {
+    slice.sort_by_key(|cc| ccc.get(u32::from(*cc)));
+}
+
+fn data_ce_to_primary(data_ce: u64, c: char) -> u32 {
+    // Collation::getThreeBytePrimaryForOffsetData
+    let p = (data_ce >> 32) as u32; // three-byte primary pppppp00
+    let lower32 = data_ce as u32 as i32; // base code point b & step s: bbbbbbss (bit 7: isCompressible)
+    let mut offset = ((u32::from(c) as i32) - (lower32 >> 8)) * (lower32 & 0x7F); // delta * increment
+    let is_compressible = (lower32 & 0x80) != 0;
+    // Collation::incThreeBytePrimaryByOffset
+    offset += (((p >> 8) & 0xFF) as i32) - 2;
+    let mut primary = (((offset % 254) + 2) as u32) << 8;
+    offset /= 254;
+    // Same with the second byte,
+    // but reserve the PRIMARY_COMPRESSION_LOW_BYTE and high byte if necessary.
+    if is_compressible {
+        offset += (((p >> 16) & 0xFF) as i32) - 4;
+        primary |= (((offset % 251) + 4) as u32) << 16;
+        offset /= 251;
+    } else {
+        offset += (((p >> 16) & 0xFF) as i32) - 2;
+        primary |= (((offset % 254) + 2) as u32) << 16;
+        offset /= 254;
+    }
+    primary | ((p & 0xFF000000) + ((offset as u32) << 24))
+}
+
+/// Iterator that transforms an iterator over `char` into an iterator
+/// over `CollationElement` with a tailoring.
+/// Not a real Rust iterator: Instead of `None` uses `NO_CE` to indicate
```

Comment:
suggestion: Exemplify this caveat through code

 Add a runnable code example in this Rust doc to exemplify this unique behavior.  I think the preference is for unit tests to be turned into Rust doc example code wherever possible.

This module doesn't have unit tests, but it probably should (or equivalent Rust doc examples).  I'm a big believer in generative / property-based testing (aka "fuzz testing"), but even I recognize that it's a supplement, not a substitute, to well-chosen thorough unit tests.  This seems like a good place to start.

Reply:
Per meeting, not adding a doc comment as this is a private API.